connector/jobrunner/runner.py

```diff
@@ -0,0 +1,307 @@
+##############################################################################
+#
+#     This file is part of connector, an Odoo module.
+#
+#     Author: Stéphane Bidoul <stephane.bidoul@acsone.eu>
+#     Copyright (c) 2015 ACSONE SA/NV (<http://acsone.eu>)
+#
+#     connector is free software: you can redistribute it and/or
+#     modify it under the terms of the GNU Affero General Public License
+#     as published by the Free Software Foundation, either version 3 of
+#     the License, or (at your option) any later version.
+#
+#     connector is distributed in the hope that it will be useful,
+#     but WITHOUT ANY WARRANTY; without even the implied warranty of
+#     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#     GNU Affero General Public License for more details.
+#
+#     You should have received a copy of the
+#     GNU Affero General Public License
+#     along with connector.
+#     If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+"""
+Odoo Connector jobs runner
+==========================
+
+What's this?
+------------
+This is an alternative to connector workers, with the goal
+of resolving issues due to the polling nature of workers:
+* jobs do not start immediately even if there is a free connector worker
+* connector workers may starve while other workers have too many jobs enqueued
+* connector workers require another startup script,
+  making deployment more difficult
+
+It is fully compatible with the connector mechanism and only
+replaces workers.
+
+How?
+----
+* It starts as a thread in the Odoo main process
+* It receives postgres NOTIFY messages each time jobs are
+  added or updated in the queue_job table.
+* It maintains an in-memory priority queue of jobs that
+  is populated from the queue_job tables in all databases.
+* It does not run jobs itself, but asks Odoo to run them through an
+  anonymous /runjob HTTP request [1].
+
+How to use
+----------
+* set the following environment variables:
+  - ODOO_CONNECTOR_RUNNER_ENABLE=1
+  - ODOO_CONNECTOR_CHANNELS=root:4 (or any other channels configuration)
+  - optional if xmlrpc_port is not set: ODOO_CONNECTOR_PORT=8069
+* start Odoo with --load=web,connector and --workers > 1 [2]
+* disable "Enqueue Jobs" cron
+* do NOT start openerp-connector-worker
+* create jobs (eg using base_import_async) and observe they
+  start immediately and in parallel
+
+TODO
+----
+* See in the code below.
+
+Notes
+-----
+[1] From a security standpoint, it is safe to have an anonymous HTTP
+    request because this request only accepts to run jobs that are
+    enqueued.
+[2] It works with the threaded Odoo server too, although this is
+    obviously not for production purposes.
+"""
+
+from contextlib import closing
+import logging
+import re
+import select
+import threading
+import time
+
+import psycopg2
+from psycopg2.extensions import ISOLATION_LEVEL_AUTOCOMMIT
+import requests
```

Comment:
> done in 8.0 branch

I reported that on the 7.0 branch because `requests` is already a dependency of Odoo 8.0. 
Anyway, it does not hurt to add in in the 8.0 branch too.